server/lib/src/idm/ldap.rs

```diff
@@ -525,6 +508,63 @@ impl LdapServer {
             },
         } // end match server op
     }
+
+    async fn bind_target_from_bind_dn(
+        &self,
+        idms: &IdmServer,
+        dn: &str,
+        pw: &str,
+    ) -> Result<LdapBindTarget, OperationError> {
+        if dn.is_empty() {
+            if pw.is_empty() {
+                return Ok(LdapBindTarget::Account(UUID_ANONYMOUS));
+            } else {
+                // This is the path to access api-token logins.
+                return Ok(LdapBindTarget::ApiToken);
+            }
+        } else if dn == "dn=token" {
+            // Is the passed dn requesting token auth?
+            // We use dn= here since these are attr=value, and dn is a phantom so it will
+            // never be present or match a real value. We also make it an ava so that clients
+            // that over-zealously validate dn syntax are happy.
+            return Ok(LdapBindTarget::ApiToken);
+        }
+
+        if let Some(captures) = self.binddnre.captures(dn) {
+            let app = captures.name("app");
+            let usr = captures.name("val");
+
+            if usr.is_some() {
+                let usr = usr.unwrap().as_str();
+
+                if usr.is_empty() {
+                    // That's weird ...
+                    return Err(OperationError::NoMatchingEntries);
+                }
+
+                let mut idm_auth = idms.auth().await;
+                let usr_uuid = idm_auth.qs_read.name_to_uuid(usr).map_err(|e| {
+                    request_error!(err = ?e, ?usr, "Error resolving rdn to target");
+                    e
+                })?;
+
+                if app.is_some() {
+                    let app = app.unwrap().as_str();
+
+                    if app.is_empty() {
+                        // That's weird ...
+                        return Err(OperationError::NoMatchingEntries);
+                    }
+
+                    return Ok(LdapBindTarget::Application(app.to_string(), usr_uuid));
+                }
+
+                return Ok(LdapBindTarget::Account(usr_uuid));
+            }
+        }
+
+        Err(OperationError::NoMatchingEntries)
```

Comment:
Error message explaining how this occured and the case.

Reply:
Done.